### ISSUE TITLE
Handling img elements

### ISSUE THREAD

**Issue by dr-coctor** [KingSora/OverlayScrollbars#16]
hello,

scroll bar (in my case horizontal, i didnt check vertical) didnt show/appearing...  only after window resize or in my case-when i activate next tab where i other img gallery) ... for test i using your css and css theme... problem come up when site is uploaded on server (or test locallhost) but when i using a brackets live view every thing is ok... on classic firefox and chrome there is no problem also (everywhere) 

in console i get massage: "This site appears to use a scroll-linked positioning effect. This may not work well with asynchronous panning; see https://developer.mozilla.org/docs/Mozilla/Performance/ScrollLinkedEffects for further details and to join the discussion on related tools and features!"

**Comment by dr-coctor** [comment]
sorry for my english.. thank you four your patience... i uploaded a beta version of my site then you can see every thing (issue is only with firefox developer edition).
in image gallery (portfolio) after first load of page there are 2 scenarios: 1. you will see a scrollbar but it will by unactive or 2. you dont see scrollbar. in those two cases you can resize window or click on link to other photo gallery...
igrok.pl

**Comment by KingSora** [comment]
I think this issue is happening because you're working with `<img />` elements.
Because `<img />` elements downloads their image which is specified in the `src`  attribute asynchronous, the plugin doesn't know whether it has to update or not. I've created a small fiddle: [here](https://jsfiddle.net/ww3oae0w/)

In this fiddle I'm simulating a slow internet connection. The image will be downloaded after 1500 milliseconds. If you take a look on the code, I've commented the fix in line 12. If you remove this line, the plugin won't update and you can't scroll.

To fix your problem, please implement a `$('img').on('load', function() { });` callback on every `<img />` element which is inside of the plugins DOM. The callback has to call the `update()` method of the plugin instance.

Please try this, I hope this will fix your problem.

**Comment by dr-coctor** [comment]
unfortunately it didn't work or (what is must likely) i did some mistakes in implementation of your fix (http://prntscr.com/is5sw4). i getting this msg ( http://prntscr.com/is5z1z )so definitely i'm doing some thing wrong...
a BIG thanks for your time and your involvement in this case... im hopeless so i fill bad for wasting your time... of corse if this issue is some kind of interesting for you then please feel free to do what ever you wont with all of data from this web page (not only for this issue but for future in other of your cases)

**Comment by KingSora** [comment]
Please change your selector from `$('.gal')` to `$('#fg_h-1')`. `$('.gal')` will give you a **array** of instances, since your webpage has more element with this class.

**Comment by dr-coctor** [comment]
it work :)) but unfortunately i have to use  ".gal". (doesn't mater for me if it . gal or ".fg_horizontal article" - this is a same container) .. as you probable notice i use "tab-idea-construct" for my img gallery and if i use id for calling plugin then others gallery containers wouldn't load with scrollbar...

please show me what i have to do, some example you mentioned...

**Comment by KingSora** [comment]
try it like this:

```javascript
var hostElementsSelector = '.gal';
//init scrollbars
$(hostElementsSelector).overlayScrollbars({ /* your options */ }).overlayScrollbars();
//update the scrollbars on img load
$(hostElementsSelector + ' img').on('load', function(e) { 
    var hostElement = $(e.currentTarget).closest(hostElementsSelector);
    //check if the host element was found
    if(hostElement) {
        var hostElementInstance = hostElement.overlayScrollbars();
        //check if the instance is initialized
        if(hostElementInstance)
            hostElementInstance.update();
    }
});

```

**Comment by dr-coctor** [comment]
A BIG thank you. its work perfect. im in shock and i dont know what to say (how to say my feelings) :)

**Comment by KingSora** [comment]
No problem, pal :)

Well, my plugin is basically native scrolling with custom scrollbars. Which means if there would be a CSS or HTML approach for this, it would work for my plugin. Unfortunately there is none. But you can use "dirty hacks" to make it possible.
I've created a small fiddle for you: https://jsfiddle.net/8at3aty9/1/
As you can see, there is no smooth scrolling but it works. Its a bit tricky if you're using a trackpad on a laptop though. You need [this](https://github.com/jquery/jquery-mousewheel) third party plugin for it to work.

**Comment by dr-coctor** [comment]
wow... i dont know what too say... thank you very much my master :) off corse i doing some thing wrong (http://prntscr.com/isoruc  |  http://prntscr.com/isosw1) because itsn working in my case.. well it doesn't matter... i'm happy and grateful for your biiiiiig work, help for me:)

**Comment by KingSora** [comment]
This should work for your site:
```javascript
var hostElementsSelector = '.gal';

//init scrollbars
var instances = $(hostElementsSelector).overlayScrollbars({ /* your options */ }).overlayScrollbars();

//update the scrollbars on img load
$(hostElementsSelector + ' img').on('load', function(e) { 
    var hostElement = $(e.currentTarget).closest(hostElementsSelector);
    //check if the host element was found
    if(hostElement) {
        var hostElementInstance = hostElement.overlayScrollbars();
        //check if the instance is initialized
        if(hostElementInstance)
            hostElementInstance.update();
    }
});

//set the horizontal mouse-scroll behavior
$.each(instances, function(index, instance) {
    var scrollValue = instance.scroll().x.position;
    var scrollDirection;
    var scrollSmoothness = 10;
    var scrollAmount = 100;
    var animationLoop = function() { 
        var currScroll = instance.scroll().x.position;
        var delta = scrollValue - currScroll;
        var easedValue = parseInt(currScroll + (delta / Math.max(1, scrollSmoothness)));
        if(easedValue === currScroll)
        easedValue += delta > 0 ? 1 : -1;
        console.log(easedValue);
        if(delta !== 0) {
            instance.scroll({ x :  easedValue });
            requestAnimationFrame(animationLoop);
        }
    };
    $(instance.getElements().viewport).mousewheel(function(event, delta) {
        if(event.originalEvent.deltaX === 0) {
            var factor = delta < 0 ? -1 : 1;
            var directionChanged = scrollDirection != factor;
            var amount = scrollAmount * factor;
            var root = directionChanged ? instance.scroll().x.position : scrollValue;
                scrollValue = Math.max(0, Math.min(instance.scroll().x.max, root - amount))
            scrollDirection = factor;
            animationLoop();
            event.preventDefault();
        }
    });
});
```

**Comment by dr-coctor** [comment]
dear gooood... again i dont know what too say, i, never get so much help from unknown person... if you dont mine i thing i should add link to your plugin web site in my footer section as a "powered by"...

**Comment by KingSora** [comment]
No problem mate, I am thanking you for your kind donation! :)
I don't mind, you can credit me where ever you like!

If you have any other issues with the plugin, don't hesitate to open a issue.

**Comment by dr-coctor** [comment]
i will not open any issue ticket because plugin is perfect ;)... a specially now :>

**Comment by KingSora** [comment]
In the newest release: https://github.com/KingSora/OverlayScrollbars/releases/tag/v1.4.1
you don't have to handle the image load events by yourself anymore. The plugin will handle this for you.
If you update to 1.4.1 you can delete the corresponding code. 
(The code which is commented with `//update the scrollbars on img load`)

**Comment by dr-coctor** [comment]
grate :) i didnt have time to even open bracket but today i checked and work "la perfecto" :)  that's cool that you decide to make this update :)
btw. i have question, just friendly question;) have you considered adding a option that changing position of scrollbar out and in box/container? or maby already it ... and i miss it in documentation... btw. your documentation is awesome one of best i have ever saw...

**Comment by KingSora** [comment]
Hehe thanks :)

Well, I've thought about that, and I am not sure if there will be ever a feature like that. 
Since scrollbars are ment to be inside a element. You can achive this effect with the `paddingAbsolute` option set to true and some clever padding values though. 

Don't get me wrong it's certainly not impossible, but it would be a greater deal to implement and I am not sure if this feature would be worth it.

**Comment by dr-coctor** [comment]
no, no ;) i totally understand you, i know it's possible (every thing is:>) but it is worthy? i was wondering so that why i asked ;)

unfortunately i have new issue and this one is critical... i wrote a simple function to load content (those img) 
```

var linki = $('.is_main-nav ul > li > ul a[href*=".txt"]');
        linki.click(function(e){
                e.preventDefault();
            var fotos = $(this).attr('href'); 
            $(".gal").load(fotos);
```
and its work but not in container with your plugin:( is there a easy fix? maybe some options build i plugin?

**Comment by KingSora** [comment]
use: `$(".gal .os-content").first().load(fotos);`
This should fix your problem I believe.

**Comment by dr-coctor** [comment]
yesss this helps ;) thank you... but wondering why, what was wrong? why this need a refresh?

**Comment by KingSora** [comment]
Its because my plugin is wrapping the content of the `.gal` container in its own container. This is necessary to make the plugin work. Here you can read the documentation: https://kingsora.github.io/OverlayScrollbars/#!documentation/domelements

**Comment by dr-coctor** [comment]
wrapping a lot :) thx

**Comment by dr-coctor** [comment]
hello :)
well i finished my site but unfortunately i find a bug... its long story but i find problem with iframe's... i dont use your scroll plugin in any iframes but i have 3 on site (2 yt, 1 google map)... scroll works grate until i put iframe on site... i did try a changing order in my calling/handling .js but it didnt work...
http://test.igrok.atthouse.pl/test-dev-eng-index.html
i didnt notice this earlier because problem show up always on ie/edge (i dont care about m$:>) and in some cases on chrome and ff... e.g my friend using this same version of chrome and os as me and he have issue with scroll and i dont...

**Comment by KingSora** [comment]
Alright, I was on your site (with Firefox) and I couldn't see anything wrong. Can you explain what exactly isn't working? Is the plugin not initializing or does just the scrolling not work? Please also check the console and tell me its content if there is any.

**Comment by dr-coctor** [comment]
thank you for your interest
i know its tricky... like i said on my pc work also fine (win10-64bit-> ff dev 60/ chrome 65/ android) but some of my friends (not all ) says that scroll not initializing and in some cases not initializing and all scripts below gallery stops working/not initializing (i made a my .js where i put my scripts and 3thparty's scripts/handlers- http://test.igrok.atthouse.pl/js/myjs-dev.js) 
i was able to recreate this issue/case only on ie/edge. this "problem" always appears on edge/ie, that why i didnt saw it earlier. scroll will not appear or not appear and all scripts below photo gallery will not initialize

**Comment by dr-coctor** [comment]
in those 2 last link's there is no any my scripts, only jq lib, scroller and classic plugin "caller" in html... but thx  for your time.. maybe some day i will figure out what was wrong :)

**Comment by KingSora** [comment]
@dr-coctor please try out [version 1.4.5](https://github.com/KingSora/OverlayScrollbars/releases/tag/v1.4.5)
I've fixed a couple of iFrame bugs, maybe your bug is also fixed.

**Comment by Ameobea** [comment]
@KingSora you're a hero
